File: NTP/ntp.cs
public static DateTime NTPTime(String TimeServer, int UTC_offset)
        {
            // Find endpoint for timeserver
            IPEndPoint ep = new IPEndPoint(Dns.GetHostEntry(TimeServer).AddressList[0], 123);

            // Connect to timeserver
            Socket s = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
            s.Connect(ep);

            // Make send/receive buffer
            byte[] ntpData = new byte[48];
            Array.Clear(ntpData, 0, 48);

            // Set protocol version
            ntpData[0] = 0x1B;

            // Send Request
            s.Send(ntpData);

            // Receive Time
            s.Receive(ntpData);

            byte offsetTransmitTime = 40;

            ulong intpart = 0;
            ulong fractpart = 0;

            for (int i = 0; i <= 3; i++)
                intpart = (intpart << 8) | ntpData[offsetTransmitTime + i];

            for (int i = 4; i <= 7; i++)
                fractpart = (fractpart << 8) | ntpData[offsetTransmitTime + i];

            ulong milliseconds = (intpart * 1000 + (fractpart * 1000) / 0x100000000L);

            s.Close();

            TimeSpan timeSpan = TimeSpan.FromTicks((long)milliseconds * TimeSpan.TicksPerMillisecond);
            DateTime dateTime = new DateTime(1900, 1, 1);
            dateTime += timeSpan;
            
            TimeSpan offsetAmount = new TimeSpan(0, UTC_offset, 0, 0, 0);
            DateTime networkDateTime = (dateTime + offsetAmount);

            return networkDateTime;
        }
		
		//call method Utility.SetLocalTime(NTPTime("time-a.nist.gov", -5));

File: NTP/ntp.vb
Public Shared Function NTPTime(TimeServer As [String], UTC_offset As Integer) As DateTime
	' Find endpoint for timeserver
	Dim ep As New IPEndPoint(Dns.GetHostEntry(TimeServer).AddressList(0), 123)

	' Connect to timeserver
	Dim s As New Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp)
	s.Connect(ep)

	' Make send/receive buffer
	Dim ntpData As Byte() = New Byte(47) {}
	Array.Clear(ntpData, 0, 48)

	' Set protocol version
	ntpData(0) = &H1b

	' Send Request
	s.Send(ntpData)

	' Receive Time
	s.Receive(ntpData)

	Dim offsetTransmitTime As Byte = 40

	Dim intpart As ULong = 0
	Dim fractpart As ULong = 0

	For i As Integer = 0 To 3
		intpart = (intpart << 8) Or ntpData(offsetTransmitTime + i)
	Next

	For i As Integer = 4 To 7
		fractpart = (fractpart << 8) Or ntpData(offsetTransmitTime + i)
	Next

	Dim milliseconds As ULong = (intpart * 1000 + (fractpart * 1000) / &H100000000L)

	s.Close()

	Dim timeSpan__1 As TimeSpan = TimeSpan.FromTicks(CLng(milliseconds) * TimeSpan.TicksPerMillisecond)
	Dim dateTime As New DateTime(1900, 1, 1)
	dateTime += timeSpan__1

	Dim offsetAmount As New TimeSpan(0, UTC_offset, 0, 0, 0)
	Dim networkDateTime As DateTime = (dateTime + offsetAmount)

	Return networkDateTime
End Function
'Utility.SetLocalTime(NTPTime("time-a.nist.gov", -5))